LIP-37: EDF (#121)
lidofinance/lido-improvement-proposals: LIPS/lip-x.md

Reviewer: fun note:

DSM signatures are not bound to a specific guardian contract. The LIP keeps the DSM signed digest unchanged and passes `guardian` alongside the signature. A single delegatee signature can validate for multiple guardian contracts if those contracts return the same `getDelegatee()`. 

Sorting by guardian only prevents duplicate `guardian` addresses, **not duplicate recovered signers**. 

With current DSM quorum 4/6 verified, this can double-count one hot key if several KDF guardians share it. Probably, the digest should include at least guardian, DSM address, chain id, action/type, and the existing message fields.

Reviewer: let's add PQ design considerations more explicitly in the text

Reviewer 2: 1. **Motivation** — `LIPS/lip-37.md` L35:
   > Future signing-scheme migration. The signing schemes KDF relies on are not post-quantum safe, and in the future the protocol might need to support such verification. Routing all signature verification through KDF means such a migration needs no protocol change, only a KDF upgrade itself.

2. **Architectural rationale** — `LIPS/lip-37.md` L66:
   > Why keep cryptography outside the protocol. All signature verification lives in KDF, so the eventual move to post-quantum signatures means updating only KDF, not the protocol. The protocol just does a generic ERC-1271 check against the delegation contract address and never needs to know how a signature is produced, so a PQ migration stays entirely within KDF.

3. **Owner implementation** — `LIPS/lip-37.md` L354:
   > Strongly recommended: Safe (Gnosis Safe) multisig. Safe is the preferred owner implementation because it is battle-tested at Lido scale and its upgrade path is expected to include post-quantum (PQ) signature support — an assumption this design explicitly relies on. A Safe-based owner should be upgradeable to PQ-compatible signing without replacing the multisig address or requiring a governance vote for the delegation contract.

Let me know if this isnt enough please

---
lip: <to be assigned>

> title: "KDF Release: HashConsensus Hardening, Timestamp-Based Frames, Key Delegation Framework, and Protocol-Wide KDF Adoption"

Reviewer 3: Quite a long title. Maybe smth like "Key Delegation Framework"?

Author: done

status: WIP
author: Raman Siamionau
discussions-to: <Create a new thread on https://research.lido.fi/ and drop the link here>
created: 2026-06-01
---

## Simple Summary

This proposal bundles four related improvements to Lido's infrastructure into a single release:

1. **Key Delegation Framework (KDF)** — introduces a general-purpose on-chain delegation layer that sits in front of any permissioned key in the protocol, allowing its operator to rotate the hot signing key instantly, without governance voting.
2. **KDF adoption within Oracle and DSM scope** — specifies the concrete off-chain and on-chain changes required to adopt KDF across Lido Oracles and Council daemons.

> 3. **HashConsensus vulnerability fix** — closes a denial-of-service vector in the `HashConsensus` contract by bounding the number of stored report variants per frame.

Reviewer 3: A link to the issue would be helpful https://github.com/lidofinance/core/issues/1379

Author: added

4. **Timestamp-based oracle frames** — replaces slot/epoch-based frame logic in `HashConsensus` with wall-clock seconds, so oracle frame lengths stay stable across future Ethereum hardforks (including the EIP-7782 slot time reduction).

## Abstract

This upgrade pursues two main goals: to improve the security of the protocol, and to abstract existing functionality away from slots and epochs so that future upgrades are simpler to perform.

> On the security side, we propose to standardize the Key Delegation Framework (KDF) as a protocol-wide primitive and to adopt it for every key that holds any protocol permission. KDF inserts a factory-deployed, per-entity delegation contract between the protocol and each operator's hot key: governance grants permissions to the delegation contract, while the operator's cold key or multisig controls which hot key is active and can rotate it at any time without a governance vote, or irreversibly lock the contract should the cold key itself be compromised. As the first batch of adopters, we propose to begin using KDF immediately for Oracle and Council operators, so that the response to a key compromise can be measured in minutes rather than the ~10 days a governance vote requires today.

Reviewer 3: > so that the response to a key compromise can be measured in minutes rather than the ~10 days a governance vote requires today.

as well as to perform regular hot key rotation

Author: Done

## Motivation

### Hot-key operational risk for Oracle and Council operators

Oracle members and Deposit Security Module Committee council members currently operate with hot EOA private keys stored directly in off-chain bots. These keys:

- Carry meaningful protocol permissions (report submission, pause/unvet signing, deposit message signing).

> - Are frequently shared across personnel and systems.

Reviewer 3: I'm not sure that's always true. It really depends on how the operator's system is set up

Author: removed

- May be long-lived with an unclear custody history.
- Require a full on-chain governance vote (~10 days) to rotate in the event of compromise.

> The 10-day exposure window between a key compromise and its on-chain revocation is an unacceptable operational risk, and the threat is not hypothetical: the May 2025 compromise of a Chorus One oracle key triggered an emergency DAO vote to rotate the affected oracle. While that incident did not exploit a specific vulnerability, it confirmed that adversarial oracle key control is a realistic scenario — and that the only remediation available today, a full governance vote, is far too slow for an active compromise.

Reviewer 3: It’s important to note that there is, after all, a quorum in place to provide protection. The long voting window itself isn’t really a major issue, and that’s not exactly what we’re addressing here.

I would frame the problem something like this: bots, by their very nature, must use hot keys. Since these keys are stored on the machine, they are easily compromised. Although the current quorum requirements cover most of the risks associated with the compromise of such accounts, the rotation process itself is still quite operationally burdensome. It requires a full-fledged vote and the involvement of the dev team and token holders in this process.

Author: done

### HashConsensus DoS vulnerability

> The current `HashConsensus` contract stores every unique report hash submitted by any oracle member in a per-frame array (`_reportVariants` / `_reportVariantsLength`). Consensus is determined by iterating the full array. Because a single oracle member can submit an unlimited number of distinct hashes, a malicious or buggy member can inflate this array, making each subsequent `submitReport` and `_isQuorumReached` call more expensive — the iteration cost grows with every additional report variant stored.

Reviewer 3: > a malicious or buggy member
I think the main vector here is key compromise and the use of the compromised key by a third party for malicious purposes.

Author: adjusted

**Why a purpose-built own delegation contract rather than reusing an existing solution.** A general-purpose delegation or smart-account framework is far more flexible than this role needs, and that unused flexibility is attack and misconfiguration surface. A minimal, non-upgradeable contract instead makes KDF's guarantees structural — admin can never `execute()` or sign, one active delegatee, instant assignment and revocation, irreversible `pause()` — keeping the audited surface small.

### Technical Specification

---

#### Part 1: Key Delegation Framework (KDF) — Contracts

##### Architecture

The delegation layer is composed of two contracts deployed from a single factory:

- **`DelegationFactory`** — a singleton that deploys `DelegationContract` instances. Each permissioned entity (oracle operator, council member) deploys exactly one `DelegationContract` per bot. A single delegation contract **must not** be shared across multiple bots or protocol roles; each independently operating bot requires its own instance.

> - **`DelegationContract`** — a non-upgradeable, minimal on-chain delegation proxy. It enforces a one-admin / one-active-delegatee model and supports two integration patterns described below.

Reviewer 3: It might be better not to use the term “delegation proxy” here so as not to confuse the reader

Author: fixed

##### Roles

The contract has exactly two roles, **admin** and **delegatee**:

| Role                    | Custody                                                    | Capabilities                                                                                                                |
|-------------------------|------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------|
| **Admin**               | Safe multisig or other audited multisig with justification | Assign, reassign, and revoke the delegatee (`assignDelegatee()` / `revokeDelegatee()`); irreversibly `pause()` the contract |
| **Delegatee (Hot Key)** | Hot key in the off-chain daemon                            | Call `execute()` to dispatch transactions (`push`); sign messages verifiable via EIP-1271 (`pull`)                          |

> The admin manages the delegatee's lifecycle but can never act *as* the contract. Pausing is the escape hatch for admin-key compromise: because `pause()` is irreversible and permanently disables `execute()` and `isValidSignature()`, a still-trusted admin can neutralise the contract entirely — stripping the delegatee of any usable role under its authority — even when the only safe move is to take the seat offline rather than rotate it.

Reviewer 4: If it is irreversible, I would not call it pause

Author: Renamed to `terminate`

> The admin manages the delegatee's lifecycle but can never act *as* the contract. Pausing is the escape hatch for admin-key compromise: because `pause()` is irreversible and permanently disables `execute()` and `isValidSignature()`, a still-trusted admin can neutralise the contract entirely — stripping the delegatee of any usable role under its authority — even when the only safe move is to take the seat offline rather than rotate it.

Reviewer 3: Should also `revokeDelegatee` in case some integrations use `getDelegatee()` for a verification

Author: 100%. Done

##### Contract Interface

```solidity
interface IDelegationContract {
    // --- Admin controls ---

    /// @notice Assign (or reassign) the active delegatee.
    ///         Only callable by admin. Takes effect immediately: the new
    ///         delegatee replaces any current one atomically, and the old
    ///         key can no longer call execute() or produce valid signatures.
    /// @param delegatee Address of the incoming delegatee.

>     function assignDelegatee(address delegatee) external;

Reviewer 3: Consider adding a cooldown period for new delegates

Author: Added.

For the councils, being offline means reduced protocol security for a while. 
For the oracles, it means no reports for some time.

so cooldown must be choosed wisely

    /// @notice Instantly remove the current delegatee. Only callable by admin.
    ///         Takes effect immediately.
    function revokeDelegatee() external;

    /// @notice Pause the contract, permanently disabling execute() and
    ///         isValidSignature(). Only callable by admin.
    ///         Intended for emergency use when the admin suspects compromise

>     ///         of the delegatee or of the contract itself.

Reviewer 3: If the Delegatee has been compromised, it is best to use `revokeDelegatee` or reassign the delegatee to a new one using `assignDelegatee`. A complete shutdown should only be used if the admin account has been compromised

Author: fixed

    ///         Pausing is irreversible; a new DelegationContract must be
    ///         deployed and governance must reassign the protocol seat.

>     function pause() external;

Reviewer 3: To be honest, the semantics are a bit unusual. A "pause" usually implies something temporary, not permanent. Consider using a different term, such as "kill" or "stop".

Author: renamed to terminate

    // --- Push integration ---

    /// @notice Execute an arbitrary call on behalf of this contract.
    ///         Only callable by the current delegatee.
    ///         Reverts if the contract is paused.
    ///         Reverts if the target call reverts, propagating the target's
    ///         revert data — the whole transaction is atomic, so forwarded
    ///         ETH is returned to the delegatee on failure.
    ///         Forwards msg.value to the target to support payable targets.
    ///         After the target call returns, any residual ETH balance on
    ///         this contract is swept back to the delegatee (msg.sender).
    /// @param target  Address to call.
    /// @param data    Call data.
    /// @return result Return data from the call.
    function execute(address target, bytes calldata data)
        external
        payable
        returns (bytes memory result);

    // --- Pull integration ---

    /// @notice EIP-1271 signature validation.

>     ///         Returns the magic value iff `signature` was produced by

Reviewer 3: ```suggestion
    ///         Returns the magic value if `signature` was produced by
```

Author: fixed

    ///         the currently assigned delegatee over `hash`.
    ///         Returns a non-magic value (does not revert) if the contract

>     ///         is paused or no delegatee is assigned.

Reviewer 3: We need just remove a delegatee if the emergency brake is pulled, so no need to check pause here

Author: fixed

    function isValidSignature(bytes32 hash, bytes calldata signature)
        external
        view

>         returns (bytes4 magicValue);

Reviewer 3: Consider looking for solutions that do not use ECDSA signatures. It would be good to have a solution that looks toward a post-quantum future.

Author: done

    // --- Views ---

    /// @notice Returns the current admin address.
    function getAdmin() external view returns (address);

    /// @notice Returns the current active delegatee, or address(0) if none.
    function getDelegatee() external view returns (address);

    /// @notice Returns true if the contract has been paused.
    function isPaused() external view returns (bool);

    // --- Events ---

    event DelegateeAssigned(address indexed newDelegatee);
    event DelegateeRevoked(address indexed revokedDelegatee);
    event Paused();
}

> interface IDelegationFactory {

Reviewer 3: Consider adding an event for the deployed instance

Author: added

    /// @notice Deploy a new DelegationContract.
    /// @param admin     Address to set as the contract's admin. Fixed for the
    ///                  lifetime of the contract — there is no on-chain way to
    ///                  change it; replacing the admin requires deploying a new
    ///                  contract and a governance vote to reassign the seat.
    /// @return instance Address of the newly deployed DelegationContract.

>     function deploy(address admin)

Reviewer 3: Consider the possibility of optionally specifying a delegate during deployment

Author: added

        │                                  │                          │
        │─ signature relayed off-chain ──────────────────────────────►│
        │                                  │                          │
        │                                  │◄──── isValidSignature() ─│
        │                                  │      (hash, sig)         │
        │                                  │                          │
        │                                  │  recover signer,         │
        │                                  │  check == getDelegatee() │
        │                                  │                          │
        │                                  │─► returns magic value ──►│
```

**Security assumptions:**
- The protocol contract must treat the `DelegationContract` address — not the hot key — as the authorized principal. It must never trust the raw signer address extracted from the signature.
- The delegatee key used for signing must correspond to the key currently stored in `getDelegatee()` at the time the signature is verified on-chain. A key rotated after signing but before verification will cause rejection.

> **When to use:** Any integration where the protocol collects signatures off-chain and verifies them on-chain in a single transaction. Examples: DSM guardian pause messages, deposit attestations, unvet signatures.

Reviewer 3: What’s missing here is that the integrator contract must know which DelegationContract to call.

From the integrator contract’s perspective, it only knows the addresses of the DelegationContracts, and the call to the contract (or the signature received) comes from the delegatee.

Possible options here:
- Cache the delegatee: iterate through each DelegationContract and cache them in a Delegatee -> DelegationContract map
- Iterate through all DelegationContracts during each validation
- The called method may contain the address of the DelegationContract (smth like `onBehalf`)

Reviewer 3: If an integrator needs to verify a signature, it can do so on its own end, and it can verify the validity of the delegation simply by calling `getDelegatee()` on DelegationContract. This way, we can keep the DelegationContract as simple as possible and free of dependencies on 1271, while achieving a PQ compatible solution

In other words, here is my proposal for the pull model:
1) delegatee call a method on an integrator contract, like `doPemissionedCall(address onBehalf)`
2) the integrator contract checks that onBehalf address is one of the privileged members
3) the integrator contract verifies delegation `onBehalf.getDelegatee() == msg.sender` (in the case of using a delegatee signature, the address recovered from the signature is used instead of msg.sender – `onBehalf.getDelegatee() == ecdsa.recover(msgHash, sig)`)

Thus, `ecdsa.recover()` will remain in the integrator contract only until it is replaced by another approach and will not require any additional changes to the KDA during migration to PQ accounts

Author: The LIP also proposes an alternative variant. The delegatee also adds the address of the delegator contract to the signature. The integrator can then directly verify that this address is a guardian address and that the signature was signed by the delegatee member of this contract.

wdyt?

Reviewer 3: Yeah, I wrote that comment while thinking about the general approach for the pull model, before I had reached the specific solution for guardians. Your solution of passing the delegator contract address is essentially the same as what I describe in option 3 (except for the place where the signature is verified). And this option seems to be the simplest and most scalable. So I agree with the proposed approach

| Validator ejector (node-operator side) | Off-chain only — must resolve the active delegatee via `getDelegatee()`. See Validator Ejector section below                 |

All components in the table are migrated together.

---

###### Oracle Daemon — Push Integration

**Pattern:** Push via `execute()`.

**Authorization model:** `HashConsensus` and the report processors (`AccountingOracle`, `ValidatorsExitBusOracle`, and CSM's `FeeOracle`) check `msg.sender` against the registered oracle committee member set. After migration, `msg.sender` is the `DelegationContract` address, which governance must have registered as the committee member before the daemon switches over.

**Daemon configuration:** `DELEGATION_CONTRACT_ADDRESS` environment variable. When set, all oracle contract calls are routed through `DelegationModule`, a Web3py extension that wraps `execute()` dispatch. Startup validation checks: (a) `getDelegatee() == configuredHotKey`, and (b) `delegationContract` address holds oracle committee membership in `HashConsensus`.

> **Transitional behavior:** When `DELEGATION_CONTRACT_ADDRESS` is unset, calls are sent directly from the hot key EOA as before. This path exists for development purposes.

Reviewer 3: > This path exists for development purposes

Also for transition and as a fallback (maybe there will be native AA delegation implementations that directly meet our needs)

Author: added

```solidity
// Per-signature payload gains the guardian address it is signed under:
struct GuardianSignature {
    address guardian;   // registered guardian: DelegationContract or legacy EOA
    bytes32 r;
    bytes32 vs;         // EIP-2098 compact signature
}

function _isValidGuardianSignature(
    address guardian,
    bytes32 msgHash,
    bytes32 r,
    bytes32 vs
) internal view returns (bool) {
    if (!_isGuardian(guardian)) return false;
    if (guardian.code.length > 0) {
        // Contract guardian: EIP-1271 over the 64-byte (2098) signature.
        // The DelegationContract recovers the signer and checks == getDelegatee().
        return IERC1271(guardian).isValidSignature(msgHash, abi.encodePacked(r, vs))
            == IERC1271.isValidSignature.selector;
    }
    // Legacy EOA guardian: recovered signer must equal the claimed guardian.
    return ECDSA.recover(msgHash, r, vs) == guardian;

> }

Reviewer 3: Consider smth like this to avoid ECDSA in DelegationContract that is incompatible with PQ

```
function _isValidGuardianSignature(
    address guardian,
    bytes32 msgHash,
    bytes32 r,
    bytes32 vs
) internal view returns (bool) {
    if (!_isGuardian(guardian)) return false;
    address recoveredAddress = ECDSA.recover(msgHash, r, vs);

    if (guardian.code.length > 0) {
        return DelegationContract(guardian).getDelegatee() == recoveredAddress;
    } else {
      return recoveredAddress == guardian;
    }
}
```

Author: done

```

For the multi-signature deposit path, sorting/dedup moves from recovered signers to the **explicit guardian addresses** — the array must be strictly ascending by `guardian`, which rejects duplicates and preserves the anti-double-count invariant:

```solidity
address prev;
for (uint256 i = 0; i < sortedGuardianSignatures.length; ++i) {
    GuardianSignature calldata gs = sortedGuardianSignatures[i];
    if (gs.guardian <= prev) revert SignaturesNotSorted();
    if (!_isValidGuardianSignature(gs.guardian, msgHash, gs.r, gs.vs)) revert InvalidSignature();
    prev = gs.guardian;

> }

Reviewer 3: Consider getting rid of ECDSA signatures in DSM and switching to onchain quorum gathering, so the contract performs no cryptography of its own and is ready to PQ scheme. Something similar to what we have in HashConsensus but with extra caution (since it is still necessary to collect transaction bundle offchain and send them via a private mempool to mitigate front-run attacks)

Reviewer: ++

```

**Affected DSM entry points:** each `Signature` / `Signature[]` argument is extended to carry the guardian address as above:

```solidity
function depositBufferedEther(
    uint256 blockNumber,
    bytes32 blockHash,
    bytes32 depositRoot,
    uint256 stakingModuleId,
    uint256 nonce,
    bytes calldata depositCalldata,
    GuardianSignature[] calldata sortedGuardianSignatures   // was Signature[]
) external;

function pauseDeposits(
    uint256 blockNumber,
    GuardianSignature calldata sig   // was Signature[]
) external;

function unvetSigningKeys(
    uint256 blockNumber,
    bytes32 blockHash,
    uint256 stakingModuleId,
    uint256 nonce,
    bytes calldata nodeOperatorIds,
    bytes calldata vettedSigningKeysCounts,
    GuardianSignature calldata sig   // was Signature[]

> ) external;

Reviewer 3: `unvetSigningKeys` and `pauseDeposits` also supports direct calls from guardians, without signature. Consider to include this flow to the spec

Author: Can we remove this possibility and force councils to send signatures included.
in case of this approach - https://github.com/lidofinance/lido-improvement-proposals/pull/121#discussion_r3390336799

Reviewer 3: I would take the opposite approach, tbh – avoid cryptography at the contract level and rely on the network

```

**Transitional behavior:** The existing ECDSA path is retained, so a guardian still registered as an EOA continues to be verified unchanged. This path exists for development purposes.

---

> ###### Council Daemon — Pull Integration (Signer Side)

Reviewer 3: How will the Data Bus work? Messages will be sent from delegates, and on the depositor bot side, they will be filtered via (DSM -> members[] -> getDelegatee())

Author: added more details to the **Required change to the depositor bot:** section

**Required change to the depositor bot:** The bot relays the council-provided `(DelegationContract address, signature)` pairs to DSM as-is. The only behavioural change is ordering: when assembling `depositBufferedEther`, the bot must sort the signature array by the council-supplied guardian address (matching DSM's new ascending-by-`guardian` ordering) rather than by recovered signer.

---

###### Validator Ejector (Node-Operator Side) — Off-Chain Update

The VEBO module of `lido-oracle` is covered by the push integration above. Separately, the **validator ejector** run by node operators — the daemon that watches `ValidatorsExitBusOracle` exit request events and sends voluntary exit messages to the consensus layer — performs no on-chain transactions and needs no `DelegationContract` of its own. It does, however, need updates to operate correctly once oracles sit behind delegation contracts:

- Any check that validates the origin of an exit request against a configured oracle address must compare against the `DelegationContract` address (the registered committee member), not the delegatee EOA.

> - Any component resolving the oracle's signing address (monitoring, dashboards, message verification) must resolve the current signer via `getDelegatee()` on the delegation contract rather than a static configured address.

Reviewer 3: > must resolve the current signer

at the time the transaction is executed, in the case of verifying the transaction signature

and there is an edge case if the rotation occurs within the same block

Author: Ejector should contain delegatee addresses hardcoded.
Fixed

Reviewer 3: > Ejector should contain delegatee addresses hardcoded.

Unfortunately, yes. And frequent rotations will only worsen the NOs UX. We should think about some alerts to automate notifications to NOs.

Hopefully, we'll be rid of Ejector soon.

---

##### Operator Migration Process

Migration is designed to be zero-downtime. All preparatory steps are completed off the critical path, leaving a single irreversible governance vote that flips every prepared seat to delegation at once. The reassignment of a committee seat from a hot EOA to a `DelegationContract` cannot be undone without a further governance vote, so operators are expected to validate the full flow on the Hoodi testnet before the mainnet vote.

1. **Deploy and configure contracts (operators)**: each operator deploys its `DelegationContract` via `DelegationFactory.deploy(adminAddress)` — the admin being a Safe multisig, fixed for the contract's lifetime (see Admin Immutability in Part 1) — and assigns the active delegatee (its existing hot key) via `assignDelegatee()`.

> 2. **Publish and verify addresses (operators)**: each operator publishes its `DelegationContract` and admin addresses on the Lido research forum, so the DAO and other operators can verify them ahead of the vote.

Reviewer 3: Should we ask for a multisig-signed message? (no need to address it in the LIP, just curious)

Author: IMO, we usually ask users to sign a message to make sure this is a real address and that the user didn’t mess it up while generating it.
I think deploying a contract from this address is enough to confirm that the address is valid, so we can skip "sign" verification.

Reviewer: there can be arbitrary calldata argument for the factory which leaves graffiti text onchain, can be used for verification of intent

4. **Set up monitoring (Lido team)**: the Lido team registers every delegation contract in the monitoring infrastructure and begins watching the admin and delegatee addresses for unexpected activity, so anomalies are caught both before and after the seat reassignment.
5. **Governance vote**: a single DAO vote reassigns each Oracle committee and DSM guardian seat from the operator's hot EOA to its `DelegationContract` address — a `HashConsensus` member update for oracles, a `DepositSecurityModule` guardian replacement for guardians; the pre-configured daemons begin routing through the delegation contract without interruption.

Until the governance vote in step 5 executes, operators continue to run as EOA principals; the EOA paths described above keep them online throughout preparation. After the vote, each operator confirms the switchover via `getDelegatee()` and a successful report or guardian message in the following frame, and may rotate the hot key at any time via `assignDelegatee()` with no further governance vote. Adoption is not optional as an end state — every Oracle and DSM operator is required to complete the migration.

---

> #### Part 3: HashConsensus — Bounded Report Variant Storage

Reviewer 3: The proposed solution (https://github.com/lidofinance/core/pull/1580/changes) quite significantly changes the implementation. 

I think this can be fixed more simply, with a targeted change in one place, something like this.

Replace this: 

```solidity
uint256 support;

if (varIndex < variantsLength) {
    support = ++_reportVariants[varIndex].support;
} else {
    support = 1;
    _reportVariants[varIndex] = ReportVariant({hash: report, support: 1});
    _reportVariantsLength = ++variantsLength;
}
```

With:

```solidity
uint256 support;

if (varIndex < variantsLength) {
    support = ++_reportVariants[varIndex].support;
} else {
    // new hash: reuse a variant slot abandoned by all members, if any.
    // This bounds the variants array length (and thus all loops over it)
    // by the total number of members.
    uint64 reuseIndex = 0;
    while (reuseIndex < variantsLength && _reportVariants[reuseIndex].support != 0) {
        ++reuseIndex;
    }
    if (reuseIndex < variantsLength) {
        varIndex = reuseIndex;
    } else {
        _reportVariantsLength = ++variantsLength;
    }
    support = 1;
    _reportVariants[varIndex] = ReportVariant({hash: report, support: 1});
}
```

Author: Thanks! 
Looks a lot better then prev solution. This version is a lot simpler to implement and reaudit. Also it's independent of the timestamp migration.
Described it in Section 3

##### Storage Change

The unbounded `_reportVariants` array is replaced with a two-dimensional mapping keyed by frame ID and member index:

```solidity
// Removed:
mapping(uint256 => ReportVariant) internal _reportVariants;
uint256 internal _reportVariantsLength;

// Added:
struct Report {
    bytes32 hash;
    uint64  refTimestamp;
}
mapping(uint256 => mapping(uint256 => Report)) internal _currentMemberHash;

> // frameId => memberIndex => Report

Reviewer 3: The `frameId` may change when the frame size is adjusted, and in theory, this could lead to collisions. Wouldn't it be better to use a timestamp?

Also, for the same `frameId`, everyone must have the same `refTimestamp` in the `Report`

So smth like that might be more efficient:
```solidity
mapping(uint256 => mapping(uint256 => bytes32)) internal _currentMemberHash;
// refTimestamp => memberIndex => reportHash
```

Author: Indeed. Fixed

```

Each oracle member has at most one stored report per frame. Old entries with a stale `refTimestamp` are ignored in consensus evaluation, eliminating the need to clear storage between frames.

##### `submitReport` Behavior

> Calling `submitReport(uint256 timestamp, bytes32 report, uint256 consensusVersion)` overwrites `_currentMemberHash[frameId][memberIndex]` unconditionally. This allows a member to correct a previously submitted hash within the same frame, which is a functional improvement over the current model.

Reviewer 3: > unconditionally

At the least, we should check whether the report has been delivered for the current frame and revert in such a scenario

> which is a functional improvement over the current model

The current implementation allows members to overwrite their reports

https://github.com/lidofinance/core/blob/master/contracts/0.8.9/oracle/HashConsensus.sol#L915-L936

Author: Got it!
Fixed

##### Consensus Evaluation (`_isQuorumReached`)

The consensus check iterates the fixed member set (bounded by committee size), fetching each member's single stored hash for the current frame and counting votes per unique hash. This iteration is O(N) in committee size, not O(N) in submitted variants.

##### Secondary Effect — Oracle Members Can Update Their Report

> As a consequence of the new model, oracle members can replace an earlier submission within the same frame. This is beneficial (allows correction of off-chain bugs mid-frame) but introduces the theoretical edge case that, if members repeatedly change their reports without converging, quorum formation could be delayed. Monitoring should alert on frames where consensus is not reached within the fast-lane window.

Reviewer 3: Same here, It already can do that.

Author: Fixed

If a member changes its hash after consensus was already reached and quorum no longer holds, `HashConsensus` calls `discardConsensusReport(refTimestamp)` on the report processor to drop the now-invalid report for that frame.

##### Applicability to CSM

> CSM's fee oracle (`src/FeeOracle.sol`, deployed as `CSFeeOracle`) is served by a separately-deployed `HashConsensus` instance built from a vendored copy of the same contract, which carries the identical unbounded `_reportVariants` storage and is therefore independently exposed to this DoS. The same bounded-storage fix is applied to CSM's `HashConsensus` in this release (see Affected Contracts in Part 4).

Reviewer 4: There were a few changes to CSM's version compared to the core one. I guess we can also incorporate them (changes are mostly cosmetic)

Author: The CSM version uses newer versions of the libraries and Solidity, but the logic is overall the same. We will definitely align them. No logic changes, so don't think we should mention it in the lip

---

#### Part 4: HashConsensus — Timestamp-Based Frames

##### Motivation

> EIP-7782 reduces Ethereum's slot time from 12 s to 6 s. All oracle frame lengths are currently expressed as `epochsPerFrame`, which translates to seconds via `epochsPerFrame × slotsPerEpoch × secondsPerSlot`. After the hardfork, the same `epochsPerFrame` value produces half the intended wall-clock duration. A seconds-based `secondsPerFrame` removes this dependency entirely.

Reviewer 3: This EIP has not been officially included in any fork; I would suggest phrasing it more cautiously. Smth like: There is a fairly high probability that Ethereum will change the slot size, as well as the epoch length, in the future (see EIPs).

Author: done

##### Interface Modifications

###### `IReportAsyncProcessor` (implemented by `BaseOracle`, and therefore by `AccountingOracle` and `ValidatorsExitBusOracle`; CSM runs a separate forked copy — see Affected Contracts below)

```solidity
// Before:
interface IReportAsyncProcessor {
    function submitConsensusReport(
        bytes32 _report,
        uint256 _refSlot,
        uint256 _deadline
    ) external;

    function discardConsensusReport(uint256 _refSlot) external;
}

// After:
interface IReportAsyncProcessor {
    function submitConsensusReport(
        bytes32 _report,
        uint256 _refTimestamp,
        uint256 _deadline
    ) external;

    function discardConsensusReport(uint256 _refTimestamp) external;
}

> ```

Reviewer 4: Use diff pls

```diff
- removed
+ added
```

Reviewer 4: Or two separate code blocks for better readibility

Author: adjusted

Existing implementations receive a `uint256` regardless; because `refTimestamp` values are numerically larger than any historical `refSlot`, all on-chain value comparisons (e.g., "must be greater than last processed value") remain valid without modification. Any logic that interprets the value semantically as a slot number is updated in this release as part of the protocol-wide ABI migration.

###### View Functions

```solidity
// Before:
function getFrameConfig() external view returns (
    uint256 initialEpoch,
    uint256 epochsPerFrame,
    uint256 fastLaneLengthSlots
);
function getCurrentFrame() external view returns (
    uint256 refSlot,
    uint256 reportProcessingDeadlineSlot
);
function getMembers() external view returns (
    address[] memory addresses,
    uint256[] memory lastReportedRefSlots
);
function getFastLaneMembers() external view returns (
    address[] memory addresses,
    uint256[] memory lastReportedRefSlots
);
function getConsensusState() external view returns (
    uint256 refSlot,
    bytes32 consensusReport,
    bool isReportProcessing
);

// After:
function getFrameConfig() external view returns (

>     uint256 initialTimestamp,

Reviewer 3: This kind of change is quite risky due to backward compatibility issues. The method signature remains the same, but the return values change. Especially scary is the migration from `initialEpoch` to `initialTimestamp`. Consider changing the method signatures, as they may be used somewhere in integrations. I think it would be better to break them explicitly than to start returning unexpected values.

Author: Yeah, we should discuss this one. For now, I added versioning to the function name.

function updateInitialEpoch(uint256 initialEpoch) external;
function setFastLaneLengthSlots(uint256 fastLaneLengthSlots) external;

// After:
function setFrameConfig(uint256 secondsPerFrame, uint256 fastLaneSeconds) external;
function submitReport(uint256 timestamp, bytes32 report, uint256 consensusVersion) external;
function updateInitialTimestamp(uint256 initialTimestamp) external;
function setFastLaneSeconds(uint256 fastLaneSeconds) external;
```

###### New Function

```solidity
function getInitialTimestamp() external view returns (uint256);
```

###### Removed Functions

```solidity
// Removed — superseded by getInitialTimestamp():
function getInitialRefSlot() external view returns (uint256);

// Removed — no longer required for frame math under seconds-based frames:
function getChainConfig() external view returns (
    uint256 slotsPerEpoch,
    uint256 secondsPerSlot,
    uint256 genesisTime
);
```

`getInitialRefSlot()` is replaced by `getInitialTimestamp()`, and `getChainConfig()` is dropped because frame math no longer depends on chain config. All in-protocol consumers of these methods are updated to the timestamp-based interface in this release.

> **Note — external consumers.** These two functions are public view methods that out-of-protocol integrations (block explorers, dashboards, indexers, analytics, third-party contracts) may read. Removing them is a breaking ABI change beyond Lido's own codebase, so before removal it requires dedicated research to enumerate external dependents and an advance announcement so integrators can move to the timestamp-based equivalents. If that research surfaces hard external dependencies that cannot migrate in time, retaining the two methods as thin read-only compatibility shims should be reconsidered rather than removing them outright.

Reviewer 4: What we can check is on-chain usage. Using Dune as you mentioned on the call. I believe this research is needed.

Author: getChainConfig | 0x852ded011285fe67063a08005c71a85690503cee | lido.AccountingOracle | 0xd624b08c83baecf0807dd2c6880c3154a5f0b288 | 4 | 2023-05-15 15:15:59 | 2023-05-15 15:15:59
-- | -- | -- | -- | -- | -- | --
getChainConfig | 0x0de4ea0184c2ad0baca7183356aea5b8d5bf5c6e | lido.ValidatorsExitBusOracle | 0x7fadb6358950c5faa66cb5eb8ee5147de3df355a | 2 | 2023-05-15 15:15:59 | 2023-05-15 15:15:59
getChainConfig | 0x4d4074628678bd302921c20573eea1ed38ddf7fb | lido.CSFeeOracle | 0x71093eff8d8599b5fa340d665ad60fa7c80688e4 | 1 | 2024-10-10 13:19:23 | 2024-10-10 13:19:23
getChainConfig | 0x442af784a788a5bd6f42a01ebe9f287a871243fb | lido.LegacyOracle | 0xd624b08c83baecf0807dd2c6880c3154a5f0b288 | 1 | 2023-05-15 15:15:59 | 2023-05-15 15:15:59
getChainConfig | 0xb2b4db1491cbe949ae85eff01e0d3ee239f110c1 |  lido.TwoPhaseFrameConfigUpdate  | 0x71093eff8d8599b5fa340d665ad60fa7c80688e4 | 1 | 2026-01-07 14:56:47 | 2026-01-07 14:56:47
getInitialRefSlot | 0x852ded011285fe67063a08005c71a85690503cee | lido.AccountingOracle | 0xd624b08c83baecf0807dd2c6880c3154a5f0b288 | 2 | 2023-05-15 15:15:59 | 2023-05-15 15:15:59
getInitialRefSlot | 0x0de4ea0184c2ad0baca7183356aea5b8d5bf5c6e | lido.ValidatorsExitBusOracle | 0x7fadb6358950c5faa66cb5eb8ee5147de3df355a | 2 | 2023-05-15 15:15:59 | 2023-05-15 15:15:59
getInitialRefSlot | 0x4d4074628678bd302921c20573eea1ed38ddf7fb | lido.CSFeeOracle | 0x71093eff8d8599b5fa340d665ad60fa7c80688e4 | 1 | 2024-10-10 13:19:23 | 2024-10-10 13:19:23

But still there is a lot of off-chain staff that could use view methods

Reviewer 4: Let's attach it to the LIP. As a hackmd link.

It is great that there are no on-chain usages other than Lido internal. Off-chain is easier to update

##### Events and Errors

All events and custom errors that reference `refSlot`, `slot`, `epoch`, `epochsPerFrame`, `fastLaneLengthSlots`, or `initialEpoch` must be updated to their timestamp-based equivalents. Off-chain services that parse these events (oracle daemons, indexers, monitoring tools) will require corresponding updates before or at the time of contract upgrading.

##### Off-Chain Oracle Daemon — Resolving `refTimestamp` to a Reference Slot

On-chain the frame is now anchored to a wall-clock `refTimestamp`, but an oracle report is still built against the consensus-layer state at a concrete **slot**. The daemon must therefore resolve the frame's `refTimestamp` to a reference slot before assembling a report:

1. **Read the frame reference.** Fetch `refTimestamp` from `HashConsensus.getCurrentFrame()`.
2. **Resolve it to a slot.** If a slot boundary falls exactly on `refTimestamp` — `(refTimestamp − genesisTime) % secondsPerSlot == 0` — that slot is the reference slot. If no slot lands exactly on `refTimestamp`, the daemon takes the **latest slot at or before `refTimestamp`** (the previous slot):

   ```

>    refSlot = (refTimestamp − genesisTime) / secondsPerSlot   // integer division (floor)

Reviewer 3: This formula is most likely incorrect, since after `genesisTime` there will be 12-second slots, and at some point, X-second slots will begin.  

```
       12s slots        8s slots    6s slots
|--------------------|------------|---------
genesis              switch       switch 
time                 to 8s slots  to 6s slots
```

It is more likely that you need to retrieve points (or just the last one) from the node's config where the slot time changes, along with their slot and timestamp, and factor them into the calculations.

_I saw below that these issues had already been addressed, but I think I'll leave the comment as it may be useful for understanding the problem_

   ```

>    The formula above assumes a single, constant slot duration (`genesisTime`/`secondsPerSlot` from the beacon-chain spec). The exact resolution logic depends on which slot-timing change is ultimately adopted and must track it: a one-off `secondsPerSlot` reduction (EIP-7782) keeps this single-constant formula but with a new value, whereas a runtime-configurable or piecewise schedule (EIP-8198) requires walking the slot-duration schedule active at `refTimestamp` rather than dividing by one constant. The invariant that must hold regardless of the adopted EIP is "resolve to the latest slot at or before `refTimestamp`"; only the arithmetic that maps timestamps to slots changes.

Reviewer 3: Does that mean they'll "add slots" retroactively?

Author: What do you mean? I didn't understand

Reviewer 3: > a one-off `secondsPerSlot` reduction (EIP-7782) keeps this single-constant formula but with a new value

I didn't quite understand the wording and wanted to clarify whether I understood correctly that you're describing here that slots will be added retroactively as soon as the constant changes?

For example, if the hard fork activation point falls on slot 100, then that slot will become slot 100 * 2 = 200, and all previous slots (< 200) will gain a “neighbor”

```
before:  0   1   2   3 ...
after:   0 1 2 3 4 5 6 ...
```

Author: Improved wording

3. **Build the entire report on that single slot.**
4. **Reference that slot in the report.** The report still carries the resolved `refSlot` so that every consumer and verifier can confirm the whole report was built against the same slot.

##### Migration Path

The new `HashConsensus` contract is initialized through a finalize-upgrade method that receives the full seconds-based frame configuration (`secondsPerFrame`, `fastLaneSeconds`). Rather than passing `initialTimestamp` in by hand, the method derives it on-chain from the legacy contract: it reads the old `getInitialRefSlot()` and `getChainConfig()` and computes

```
initialTimestamp = genesisTime + initialRefSlot × secondsPerSlot
```

i.e. the wall-clock timestamp of the legacy initial reference slot. Deriving the anchor on-chain from the contract being replaced guarantees the timestamp-based frame boundary lines up exactly with the legacy slot-based boundary, eliminating the risk that an off-chain miscalculation introduces a reporting gap or overlap at the migration point. The `BaseOracle` initialization flow that currently calls `getInitialRefSlot()` will call `getInitialTimestamp()` instead.

> ##### Affected Contracts and Migration Scope

Reviewer 4: What about DG?

Author: Dual Governance references `ValidatorsExitBusOracle` only by address, reading its pause state as a registered sealable (tiebreaker sealable withdrawal blocker). 
Since VEBO is under proxy and pause is unaffected - we safe here

Reviewer 4: Will you add a note on that to the LIP?

|---------------------------------------------|------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `HashConsensus.sol` (CSM)                   | Upgrade required | Identical epoch/slot frame math **and** the same unbounded `_reportVariants` storage — so CSM's fee oracle is independently exposed to the same DoS. Both fixes apply here too. |
| `BaseOracle.sol` (CSM)                      | Upgrade required | Same `getChainConfig()` / `getInitialRefSlot()` calls and `ConsensusReport.refSlot` as core's `BaseOracle`.                                                                     |
| `FeeOracle.sol` (deployed as `CSFeeOracle`) | Rename           | Passes `data.refSlot` through only (no slot arithmetic) — easier than `core`'s `AccountingOracle`.                                                                              |

> | `TwoPhaseFrameConfigUpdate.sol`             | To remove        | Frame-config update utility tied to the old slot-based `setFrameConfig`.                                                                                                        |

Reviewer 4: Not to remove, but to update for the future usage

Author: fixed

---

> lip: <to be assigned>

Reviewer: ```suggestion
lip: 37
```

title: "Key Delegation Authority"
status: WIP
author: Raman Siamionau
discussions-to: <Create a new thread on https://research.lido.fi/ and drop the link here>
created: 2026-06-01
---

> ## Simple Summary

Reviewer: The summary doesn't quite align with the title. I'd suggest one of two directions:
- Narrow the summary to focus solely on the key delegation framework, or
- Reframe the entire LIP more broadly as an improvement to the resilience of the protocol's off-chain components.

Either works — the main thing is that the title and summary tell the same story.

Reviewer: maybe we better split this into two lips tbh

This proposal bundles four related improvements to Lido's infrastructure into a single release:

> 1. **Key Delegation Framework (KDF)** — introduces a general-purpose on-chain delegation layer that sits in front of any permissioned key in the protocol, allowing its operator to rotate the hot signing key instantly, without governance voting.

Reviewer: not any please, this explanation reads weird a bit

Reviewer: I'd rather shape KDF as a mechanism design behind delegations recommended by the protocol off-chain components and broader usage

Author: done

> 1. **Key Delegation Framework (KDF)** — introduces a general-purpose on-chain delegation layer that sits in front of any permissioned key in the protocol, allowing its operator to rotate the hot signing key instantly, without governance voting.

Reviewer: is it possible to "rotate the hot signing key" instantly tho? I guess not due to cooldown (which is good)

> 2. **KDF adoption within Oracle and DSM scope** — specifies the concrete off-chain and on-chain changes required to adopt KDF across Lido Oracles and Council daemons.

Reviewer: ```suggestion
2. **KDF adoption within Oracle and DSM scope** — specifies the concrete off-chain and on-chain changes required to adopt KDF across [Lido Oracle](https://docs.lido.fi/holders/lido-oracle) and [Lido Council Daemon](https://docs.lido.fi/holders/lido-council-daemon).
```

Author: done

> 3. **HashConsensus vulnerability fix** — closes a denial-of-service vector in the `HashConsensus` contract by bounding the number of stored report variants per frame ([`lidofinance/core#1379`](https://github.com/lidofinance/core/issues/1379)).

Reviewer: ```suggestion
3. **HashConsensus weakness fix** — closes the known denial-of-service vector in the `HashConsensus` contract by bounding the number of stored report variants per frame ([`lidofinance/core#1379`](https://github.com/lidofinance/core/issues/1379)).
```

Author: done

> 4. **Timestamp-based oracle frames** — replaces slot/epoch-based frame logic in `HashConsensus` with wall-clock seconds, so oracle frame lengths stay stable across future Ethereum hardforks (including the EIP-7782 slot time reduction).

Reviewer: ```suggestion
4. **Timestamp-based oracle frames** — replaces slot/epoch-based frame logic in `HashConsensus` with wall-clock seconds, so oracle frame lengths stay stable across future Ethereum hardforks allowing dynamic slot duration (including the [EIP-8192](https://eips.ethereum.org/EIPS/eip-8198) quick slots support).
```

Author: done

> ## Abstract

Reviewer: this abstract doesn't try to explain why we batch these two independent topics together, and I wouldn't do it 'for the sake of delivery', but rather try to explain that it has shared scope and locality of the upgrade, or smth like this

> This upgrade pursues two main goals: to improve the security of the protocol, and to abstract existing functionality away from slots and epochs so that future upgrades are simpler to perform.

Reviewer: reads a bit artificial to combine these topics in such a way

Author: splitted

> On the security side, we propose to standardize the Key Delegation Framework (KDF) as a protocol-wide primitive and to adopt it for every key that holds any protocol permission. KDF inserts a factory-deployed, per-entity delegation contract between the protocol and each operator's hot key: governance grants permissions to the delegation contract, while the operator's cold key or multisig controls which hot key is active and can rotate it at any time without a governance vote, or irreversibly lock the contract should the cold key itself be compromised. As the first batch of adopters, we propose to begin using KDF immediately for Oracle and Council operators, so that the response to a key compromise can be measured in minutes rather than the ~10 days a governance vote requires today. Beyond incident response, periodic hot-key rotation is itself a valuable security practice, and KDF makes it possible to perform it routinely.

Reviewer: As wrote above, I'd suggest formulating KDF as mechanism that allows delegation suitable for permissioned actors used for the off-chain parts of the protocol and possible broader use cases

Author: done

> On the abstraction side, we propose to rewrite the `HashConsensus` contract ([`lidofinance/core`](https://github.com/lidofinance/core)) to express all frame logic in wall-clock seconds (`secondsPerFrame`, `fastLaneSeconds`, `initialTimestamp`, `refTimestamp`) instead of epoch- and slot-based parameters. This decouples oracle frame durations from Ethereum's consensus-layer slot timing and future-proofs oracle operation against changes such as the [EIP-7782](https://eips.ethereum.org/EIPS/eip-7782) slot-time reduction. The same redeployment also closes a denial-of-service vector in `HashConsensus`: a member can no longer inflate the report-variant array with distinct hashes, because abandoned variant slots are reused, capping the array — and every loop over it — at the committee size, eliminating the unbounded iteration that enables the attack.

Reviewer: pls change eip (as above)

Author: done

> On the abstraction side, we propose to rewrite the `HashConsensus` contract ([`lidofinance/core`](https://github.com/lidofinance/core)) to express all frame logic in wall-clock seconds (`secondsPerFrame`, `fastLaneSeconds`, `initialTimestamp`, `refTimestamp`) instead of epoch- and slot-based parameters. This decouples oracle frame durations from Ethereum's consensus-layer slot timing and future-proofs oracle operation against changes such as the [EIP-7782](https://eips.ethereum.org/EIPS/eip-7782) slot-time reduction. The same redeployment also closes a denial-of-service vector in `HashConsensus`: a member can no longer inflate the report-variant array with distinct hashes, because abandoned variant slots are reused, capping the array — and every loop over it — at the committee size, eliminating the unbounded iteration that enables the attack.

Reviewer: pls change eip (as above)

Concretely, we propose to deploy an updated `HashConsensus` contract, a new `DelegationFactory` / `DelegationContract` system (repository: [`lidofinance/delegation-execution-authority`](https://github.com/lidofinance/delegation-execution-authority)), and targeted updates to the [`lido-oracle`](https://github.com/lidofinance/lido-oracle) daemon and `DepositSecurityModule`. The adoption scope specifies the integration points: the `lido-oracle` daemon gains delegation support via a `DELEGATION_CONTRACT_ADDRESS` environment variable, and the `DepositSecurityModule` is updated to verify guardian signatures by recovering the signer and checking it against the guardian's `DelegationContract.getDelegatee()`, enabling council members to operate behind delegation contracts. Adoption is mandatory for these roles; a coordinated migration process is defined for existing Oracle and DSM operators to onboard to the new model without service interruption.

> ## Motivation

Reviewer: can add some short blurb on the recent incidents about leaked keys and compromised machines (>80% of the DeFi security incidents in 2026)

Author: done

### Hot-key operational risk for Oracle and Council operators

Oracle members and Deposit Security Module Committee council members currently operate with hot EOA private keys stored directly in off-chain bots. These keys:

> - Carry meaningful protocol permissions (report submission, pause/unvet signing, deposit message signing).

Reviewer: ```suggestion
- Carry meaningful protocol off-chain components permissions (report submission, pause/unvet signing, deposit message signing).
```

Author: Fixed

- May be long-lived with an unclear custody history.

> - Require a full on-chain governance vote (~10 days) to rotate in the event of compromise.

Reviewer: ```suggestion 
- Require a full on-chain governance vote (~10 days) to rotate in the event of compromise or out of abundance precaution measures. 
```

Author: fixed

> By their very nature, these bots must operate with hot keys: the signing key lives on the machine, which makes it inherently exposed and comparatively easy to compromise. The quorum requirements these roles operate under already absorb most of the risk of any single such key being compromised — a lone compromised member cannot, on its own, move the protocol — so the long voting window is not in itself the core concern. The real problem is operational: rotating a hot key today requires a full on-chain governance vote, drawing in the dev team and token holders for what should be a routine key-management action. That makes rotation burdensome both for proactive, periodic rotation and, more acutely, for responding to an active compromise. And because a replacement key is only authorized once the vote executes, an operator that stops trusting a suspect key cannot use a new one in the meantime — the seat simply stops participating until governance acts.

Reviewer: Sounds a bit weird 🤔 

Motivation now partly argues against its own urgency. Here we state "the long voting window is not in itself the core concern" and "a lone compromised member cannot, on its own, move the protocol", while the Abstract (L23) still sells "response … in minutes rather than the ~10 days". Both can be reconciled (the real benefit is routine rotation, not incident speed), but as written the two sections pull in opposite directions. 

Pick one narrative please.

> ### HashConsensus DoS vulnerability

Reviewer: ```suggestion
### HashConsensus DoS weakness
```

Author: Done

The attack works by submitting so many alternative report variants that iterating over them no longer fits within a single transaction's gas limit. At that point report submission can no longer be completed in one transaction and quorum can no longer be reached. The practical consequence is a halt of oracle reporting: for as long as the attack is sustained, `HashConsensus` cannot reach consensus and the protocol cannot process the accounting reports that drive stETH rebases, withdrawal request finalization, and validator exit requests. The DoS therefore degrades core protocol operation, not merely the `HashConsensus` contract in isolation.

> ### Ethereum slot-time changes (EIP-7782, EIP-8198)

Reviewer: would bet on 8198 only

> Ethereum's consensus-layer slot timing is in flux. Competing proposals such as [EIP-7782](https://eips.ethereum.org/EIPS/eip-7782) (12 s → 6 s) and [EIP-8198](https://eips.ethereum.org/EIPS/eip-8198) ("Quick Slots", 12 s → 8 s) would reduce slot duration and, consequently, halve or otherwise shorten the duration of all oracle frames currently configured in epochs. Under EIP-7782, for example, oracle reporting periods would shrink from roughly 28 days to 14 days, disrupting downstream protocol assumptions.

Reviewer: ```suggestion
Ethereum's consensus-layer slot timing is in flux. Competing proposals such as [EIP-7782](https://eips.ethereum.org/EIPS/eip-7782) (12 s → 6 s) and [EIP-8198](https://eips.ethereum.org/EIPS/eip-8198) ("Quick Slots", 12 s → 8 s) would reduce slot duration and, consequently, halve or otherwise shorten the duration of all oracle frames currently configured in epochs. Under EIP-7782, for example, performance oracle reporting periods would shrink from roughly 28 days to 14 days, disrupting downstream protocol assumptions.
```

EIP-8198 goes further by making slot duration a runtime-configurable parameter rather than a hardcoded constant, with the value expected to be tuned iteratively across devnets. Any frame logic anchored to `slotsPerEpoch × secondsPerSlot` is therefore not just fragile across a single hardfork but perpetually exposed to re-tuning. Migrating frame configuration from `epochsPerFrame` to `secondsPerFrame` removes the slot dependency entirely, making frame duration invariant to any present or future slot-timing change.

> ## Specification

Reviewer: I would describe KDF first without oracle and DSM specifics

Reviewer: so like in simple summary: mechanism and general-purpose design first, application for oracle and dsm — second

Author: done

### Overview

The release consists of four coordinated changes:

1. Deployment of the `DelegationFactory`, establishing the general-purpose on-chain delegation infrastructure (KDF).
2. An update to the `DepositSecurityModule` contract to verify guardian signatures against the guardian's `DelegationContract.getDelegatee()`, so that Council guardians can operate behind KDF delegation contracts.
3. Modifications to the Oracle and Council daemons to support operating behind delegation contracts.
4. Modifications to the `HashConsensus` contract and related contracts — in both `lidofinance/core` and CSM's separately-deployed fork — to fix the DoS vulnerability and switch to seconds-based frames.

### Rationale

> **Why the changes ship as one release.** The proposal groups two otherwise independent tracks — the KDF security work and the `HashConsensus` rework — into a single release to amortize the audit, deployment, and operator-coordination effort.

Reviewer: should go much earlier

### Technical Specification

---

#### Part 1: Key Delegation Framework (KDF) — Contracts

##### Architecture

The delegation layer is composed of two contracts deployed from a single factory:

- **`DelegationFactory`** — a singleton that deploys `DelegationContract` instances. Each permissioned entity (oracle operator, council member) deploys exactly one `DelegationContract` per bot. A single delegation contract **must not** be shared across multiple bots or protocol roles; each independently operating bot requires its own instance.
- **`DelegationContract`** — a non-upgradeable, minimal on-chain delegation contract. It enforces a one-admin / one-active-delegatee model and supports two integration patterns described below.

##### Roles

> The contract has exactly two roles, **admin** and **delegatee**:

Reviewer: ```suggestion
The contract has exactly two trusted entities, **admin** and **delegatee**:
```

    // --- Admin controls ---

    /// @notice Assign (or reassign) the active delegatee.
    ///         Only callable by admin. The previous delegatee is dropped at
    ///         once and the new one becomes effective only after the contract's
    ///         cooldown (`getCooldown()` seconds).
    ///         Reverts if delegatee == admin.
    /// @param delegatee Address of the incoming delegatee.
    function assignDelegatee(address delegatee) external;

    /// @notice Remove the current and any pending delegatee. 
    ///         Only callable by admin.
    function revokeDelegatee() external;

    /// @notice Terminate the contract, permanently disabling execute().
    ///         Only callable by admin.
    ///         Also clears the active delegatee (as revokeDelegatee), so
    ///         getDelegatee() returns address(0) after termination.
    ///         Intended for emergency use when the admin is suspected compromised.
    ///         Termination is irreversible; a new DelegationContract must be
    ///         deployed and governance must reassign the protocol seat.
    function terminate() external;

    // --- Push integration ---

>     /// @notice Execute an arbitrary call on behalf of this contract.

Reviewer: ```suggestion
    /// @notice Execute an arbitrary non-delegate call on behalf of this contract.
```

    /// @param cooldown  Seconds a reassigned delegatee waits before becoming
    ///                  effective (see assignDelegatee). Set in the constructor
    ///                  and immutable thereafter; may be 0 to disable the
    ///                  cooldown.
    /// @return instance Address of the newly deployed DelegationContract.
    function deploy(address admin, address delegatee, uint256 cooldown)
        external
        returns (address instance);

    /// @notice Emitted for each DelegationContract deployed by the factory,
    ///         so the full set of KDF instances can be discovered on-chain
    ///         (e.g. for the monitoring registry) by indexing this event.
    event DelegationContractDeployed(
        address indexed instance,
        address indexed admin,
        address delegatee
    );
}
```

##### Integration Models

The `DelegationContract` is designed as a **general integration primitive** for any permissioned bot in the Lido ecosystem. Two complementary patterns are supported; protocol integrations should select the one that fits their verification model.

###### Pull Style — Signature Verification against `getDelegatee()`

> The integrator (protocol) contract verifies a message itself: it recovers the signer from the signature and checks that it matches the registered entity's current delegatee, read via `getDelegatee()`.

Reviewer: prob should support [ERC-165: Standard Interface Detection](https://eips.ethereum.org/EIPS/eip-165) then

Reviewer 2: done

        │                                  │                           │
        │                                  │◄──── getDelegatee() ──────│
        │                                  │───── delegatee ──────────►│
        │                                  │                           │
        │                                  │   integrator checks:      │
        │                                  │   ecrecover(sig)==        │
        │                                  │   delegatee, and addr is  │
        │                                  │   a registered principal  │
```

**Security assumptions:**
- The integrator must treat the `DelegationContract` address — not the recovered hot key — as the authorized principal: it checks that the supplied `DelegationContract` is a registered entity (e.g. guardian) and that the recovered signer equals that contract's `getDelegatee()`.
- The signing key must match `getDelegatee()` at the time of on-chain verification. A key rotated (or the contract terminated, which sets `getDelegatee()` to `address(0)`) after signing but before verification causes rejection — failing closed.

> - The integrator **must reject a zero delegatee.** `ecrecover` returns `address(0)` for a malformed signature, and `getDelegatee()` returns `address(0)` when no delegatee is assigned (revoked/terminated/unassigned). Comparing the two naively would let a garbage signature validate as `0 == 0`, so the integrator must require `getDelegatee() != address(0)` (equivalently, reject a recovered `address(0)`) before accepting.

Reviewer: would not use bare ecrecover, but rather implement with isValidSignature, jic

https://docs.openzeppelin.com/contracts/5.x/accounts#signature-validation

Reviewer 2: done

+    uint256 initialTimestamp,
+    uint256 secondsPerFrame,
+    uint256 fastLaneSeconds
+);
-function getCurrentFrame() external view returns (
-    uint256 refSlot,
-    uint256 reportProcessingDeadlineSlot
-);
+function getCurrentFrameV2() external view returns (
+    uint256 refTimestamp,
+    uint256 reportProcessingDeadlineTimestamp
+);
-function getMembers() external view returns (
-    address[] memory addresses,
-    uint256[] memory lastReportedRefSlots
-);
+function getMembersV2() external view returns (
+    address[] memory addresses,
+    uint256[] memory lastReportedRefTimestamps
+);
-function getFastLaneMembers() external view returns (
-    address[] memory addresses,
-    uint256[] memory lastReportedRefSlots
-);
+function getFastLaneMembersV2() external view returns (
+    address[] memory addresses,
+    uint256[] memory lastReportedRefTimestamps
+);
-function getConsensusState() external view returns (
-    uint256 refSlot,
-    bytes32 consensusReport,
-    bool isReportProcessing
-);
+function getConsensusStateV2() external view returns (
+    uint256 refTimestamp,
+    bytes32 consensusReport,
+    bool isReportProcessing
+);
```

> **Why the `V2` suffix.** Each of these view getters changes its return **semantics** from slots/epochs to timestamps, so it is given a `V2` name to change its selector. If the name were kept, the selector would be identical (the getters take no arguments) and an out-of-protocol caller of e.g. `getFrameConfig()` would keep compiling and silently receive a timestamp where it expected an epoch — the most dangerous outcome.

Reviewer: the state-changing `setFrameConfig(uint256,uint256)` and `submitReport(uint256,bytes32,uint256)` keep their exact selectors despite changing `epochsPerFrame→secondsPerFrame` and `slot→timestamp`. 

These are ofc permissioned/in-protocol callers (lower risk), but the doc should say so explicitly rather than leaving the principle half-applied — `submitReport` in particular is the most-simulated function.

| `AccountingOracle.sol`                               | Upgrade required                               | Interprets the reference as a slot: the `_getSlotTimestamp(slot)` helper computes `GENESIS_TIME + slot * SECONDS_PER_SLOT`, and `timeElapsed = (data.refSlot − prevRefSlot) * SECONDS_PER_SLOT`. With a timestamp reference these conversions must change (the value is already a timestamp / a second-delta), or `onOracleReport` and exit-time math silently corrupt. |
| `ValidatorsExitBusOracle.sol`                        | Rename                                         | `DataProcessingState.refSlot` + an equality check; opaque pass-through.                                                                                                                                                                                                                                                                                                 |

> | `RefSlotCache.sol`, `VaultHub.sol`, `LazyOracle.sol` | Upgrade required                               | Use the frame reference purely as an opaque `uint48` cache key (no slot arithmetic), so they are functionally unaffected — but the call site must move from the removed `getCurrentFrame()` to `getCurrentFrameV2()`, so they must be updated/redeployed. The cached value simply becomes a timestamp. `ILazyOracle._vaultsDataRefSlot` is a parameter rename.          |

Reviewer: contradicts with L719 below

> | `OracleReportSanityChecker.sol`                      | Upgrade required (for ZK-oracle compatibility) | Uses `refSlot` to align the ZK-oracle report with the accounting report during a negative rebase. Only needed if the ZK oracle is connected to the protocol; otherwise the change can be skipped.                                                                                                                                                                       |

Reviewer: not zk oracle exactly, but rather a second opinion

> **`lidofinance/community-staking-module` (CSM):** CSM does **not** depend on `core`'s `HashConsensus`; it vendors its own copies under `src/lib/base-oracle/` and runs a **separate deployed `HashConsensus` instance** for its fee oracle (`src/FeeOracle.sol`). That fork carries the same two problems and must be migrated in lock-step (own PR, own deployment):

Reviewer: 301 → lidofinance/staking-modules

| `DelegationFactory`                | DAO / deployer                 | Single new deployment establishing the KDF infrastructure.                                                               |
| `DelegationContract` (one per bot) | each operator, via the factory | One instance per Oracle/Council bot; deployed during the operator migration (Part 2), not by a single governance action. |

**Redeployed at a new address (standalone, non-upgradeable) — references must be repointed by governance:**

| Contract                        | Repointing required                                                                                                            |
|---------------------------------|--------------------------------------------------------------------------------------------------------------------------------|
| `HashConsensus` (core)          | Set as the consensus contract on `AccountingOracle` and `ValidatorsExitBusOracle` (`setConsensusContract`).                    |
| `HashConsensus` (CSM, vendored) | Set as the consensus contract on CSM's `FeeOracle`.                                                                            |

> | `DepositSecurityModule`         | Update the `LidoLocator` reference and re-grant its deposit role; update any monitoring/config that hardcodes the DSM address. |

Reviewer: I am not sure why role re-grant?

https://github.com/lidofinance/core/blob/master/contracts/0.4.24/Lido.sol#L641

The delegation model shifts trust from "this hot key is the oracle" to "this delegation contract is the oracle." The delegation contract is non-upgradeable and minimal in scope. The admin Safe multisig is the new security boundary.

**Admin cannot act as bot by design.** The admin address has no ability to call `execute()`, and is never the delegatee returned by `getDelegatee()`, so it cannot produce signatures any integrator will accept. This is enforced at the contract level: the admin is kept maximally cold and can only manage the delegatee lifecycle.

**Pull integration (`getDelegatee()`):** Integrator contracts must treat the `DelegationContract` address as the authorized principal — checking both that it is a registered entity and that the recovered signer equals its `getDelegatee()`.

**Push integration (`execute()`):** The delegatee can call any target. Permission grants must be narrowly scoped per contract, enforced by the one-contract-per-bot rule. ETH is forwarded to the target and any change is swept back to the delegatee after the call, so overpayment refunded by the target is recovered; only ETH the target actually consumes is non-recoverable.

> **Admin compromise.** An attacker with admin access can assign a delegatee they control, but it cannot act until the `cooldown` elapses — and the `DelegateeAssigned` event (with its `activeFrom`) and `getPendingDelegatee()` expose the pending assignment on-chain for the whole window. This gives the DAO and monitoring a reaction window before the malicious key becomes effective: the legitimate admin can `terminate()`. The attacker also cannot call `execute()` or sign directly (the admin-cannot-execute invariant holds), so they must route actions through an assigned delegatee, and the blast radius is bounded to that single contract's permissions. With a zero cooldown there is no such window.

Reviewer: zero cooldown is dangerous in other way around tho

## Links

- EIP-7782 (Ethereum slot time reduction): https://eips.ethereum.org/EIPS/eip-7782
- EIP-8198 (Quick Slots — runtime-configurable slot duration): https://eips.ethereum.org/EIPS/eip-8198
- Core repository (`lidofinance/core`): https://github.com/lidofinance/core
- Delegation contracts repository (`lidofinance/delegation-execution-authority`): https://github.com/lidofinance/delegation-execution-authority
- Oracle daemon (`lidofinance/lido-oracle`): https://github.com/lidofinance/lido-oracle
- Community Staking Module — forked oracle stack (`lidofinance/community-staking-module`): https://github.com/lidofinance/community-staking-module
- HashConsensus DoS vulnerability issue (`lidofinance/core#1379`): https://github.com/lidofinance/core/issues/1379

---

```my python
                           (o)(o)
                          /     \
                         /       |
                        /   \ .. |
          ________     /    /\__/
  _      /        \   /    /   \\
 / \    /  ____    \_/    /
//\ \  /  /    \         /
V  \ \/  /      \       /
    \___/        \_____/

> ```

Reviewer 3: 🐍
